File: duplicates-android/app/src/main/java/com/github/duplicates/call/CallLogViewHolder.java
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.call;

import android.content.Context;
import android.text.format.DateUtils;
import android.view.View;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.DrawableRes;
import androidx.annotation.StringRes;

import com.github.android.removeduplicates.BuildConfig;
import com.github.android.removeduplicates.R;
import com.github.duplicates.DuplicateItemPair;
import com.github.duplicates.DuplicateViewHolder;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.github.duplicates.call.CallLogComparator.DATE;
import static com.github.duplicates.call.CallLogComparator.DURATION;
import static com.github.duplicates.call.CallLogComparator.NAME;
import static com.github.duplicates.call.CallLogComparator.NUMBER;
import static com.github.duplicates.call.CallLogComparator.TYPE;
import static com.github.duplicates.call.CallLogItem.ANSWERED_EXTERNALLY_TYPE;
import static com.github.duplicates.call.CallLogItem.BLOCKED_TYPE;
import static com.github.duplicates.call.CallLogItem.INCOMING_TYPE;
import static com.github.duplicates.call.CallLogItem.MISSED_TYPE;
import static com.github.duplicates.call.CallLogItem.OUTGOING_TYPE;
import static com.github.duplicates.call.CallLogItem.REJECTED_TYPE;
import static com.github.duplicates.call.CallLogItem.VOICEMAIL_TYPE;

/**
 * View holder of a duplicate call.
 *
 * @author moshe.w
 */
public class CallLogViewHolder extends DuplicateViewHolder<CallLogItem> {

    @BindView(R.id.match)
    TextView match;

    @BindView(R.id.checkbox1)
    CheckBox checkbox1;
    @BindView(R.id.date1)
    TextView date1;
    @BindView(R.id.duration1)
    TextView duration1;
    @BindView(R.id.number1)
    TextView number1;
    @BindView(R.id.type1)
    ImageView type1;
    @BindView(R.id.name1)
    TextView name1;

    @BindView(R.id.checkbox2)
    CheckBox checkbox2;
    @BindView(R.id.date2)
    TextView date2;
    @BindView(R.id.duration2)
    TextView duration2;
    @BindView(R.id.number2)
    TextView number2;
    @BindView(R.id.type2)
    ImageView type2;
    @BindView(R.id.name2)
    TextView name2;

    public CallLogViewHolder(View itemView) {
        this(itemView, null);
    }

    public CallLogViewHolder(View itemView, OnItemCheckedChangeListener onCheckedChangeListener) {
        super(itemView, onCheckedChangeListener);
        ButterKnife.bind(this, itemView);
    }

    @Override
    protected void bindHeader(Context context, DuplicateItemPair<CallLogItem> pair) {
        match.setText(context.getString(R.string.match, percentFormatter.format(pair.getMatch())));
    }

    @Override
    protected void bindItem1(Context context, CallLogItem item) {
        checkbox1.setChecked(item.isChecked());
        checkbox1.setText(BuildConfig.DEBUG ? Long.toString(item.getId()) : "");
        date1.setText(DateUtils.formatDateTime(context, item.getDate(), DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_SHOW_TIME | DateUtils.FORMAT_ABBREV_ALL));
        duration1.setText(DateUtils.formatElapsedTime(item.getDuration()));
        type1.setImageResource(getTypeIcon(item.getType()));
        type1.setContentDescription(context.getText(getTypeName(item.getType())));
        number1.setText(item.getNumber());
        name1.setText(item.getName());
    }

    @Override
    protected void bindItem2(Context context, CallLogItem item) {
        checkbox2.setChecked(item.isChecked());
        checkbox2.setText(BuildConfig.DEBUG ? Long.toString(item.getId()) : "");
        date2.setText(DateUtils.formatDateTime(context, item.getDate(), DateUtils.FORMAT_SHOW_DATE | DateUtils.FORMAT_SHOW_TIME | DateUtils.FORMAT_ABBREV_ALL));
        duration2.setText(DateUtils.formatElapsedTime(item.getDuration()));
        type2.setImageResource(getTypeIcon(item.getType()));
        type2.setContentDescription(context.getText(getTypeName(item.getType())));
        number2.setText(item.getNumber());
        name2.setText(item.getName());
    }

    @Override
    protected void bindDifference(Context context, DuplicateItemPair<CallLogItem> pair) {
        boolean[] difference = pair.getDifference();

        bindDifference(date1, date2, difference[DATE]);
        bindDifference(duration1, duration2, difference[DURATION]);
        bindDifference(type1, type2, difference[TYPE]);
        bindDifference(number1, number2, difference[NUMBER]);
        bindDifference(name1, name2, difference[NAME]);
    }

    @DrawableRes
    private int getTypeIcon(int type) {
        switch (type) {
            case INCOMING_TYPE:
                return R.drawable.ic_call_received_black;
            case OUTGOING_TYPE:
                return R.drawable.ic_call_made_black;
            case MISSED_TYPE:
                return R.drawable.ic_call_missed_black;
            case VOICEMAIL_TYPE:
                return R.drawable.ic_voicemail_black;
            case REJECTED_TYPE:
                return R.drawable.ic_cancel_black;
            case BLOCKED_TYPE:
                return R.drawable.ic_block_black;
            case ANSWERED_EXTERNALLY_TYPE:
                return R.drawable.ic_devices_other_black;
            default:
                return R.drawable.ic_call_black;
        }
    }

    @StringRes
    private int getTypeName(int type) {
        switch (type) {
            case INCOMING_TYPE:
                return R.string.call_type_incoming;
            case OUTGOING_TYPE:
                return R.string.call_type_outgoing;
            case MISSED_TYPE:
                return R.string.call_type_missed;
            case VOICEMAIL_TYPE:
                return R.string.call_type_voicemail;
            case REJECTED_TYPE:
                return R.string.call_type_rejected;
            case BLOCKED_TYPE:
                return R.string.call_type_blocked;
            case ANSWERED_EXTERNALLY_TYPE:
                return R.string.call_type_external;
            default:
                return R.string.call_type_other;
        }
    }

    @OnClick(R.id.checkbox1)
    void checkbox1Clicked() {
        if (onCheckedChangeListener != null) {
            onCheckedChangeListener.onItemCheckedChangeListener(item1, checkbox1.isChecked());
        }
    }

    @OnClick(R.id.checkbox2)
    void checkbox2Clicked() {
        if (onCheckedChangeListener != null) {
            onCheckedChangeListener.onItemCheckedChangeListener(item2, checkbox2.isChecked());
        }
    }
}

File: duplicates-android/app/src/main/java/com/github/duplicates/call/CallLogViewHolder.kt
/*
 * Copyright 2016, Moshe Waisberg
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.duplicates.call

import android.content.Context
import android.text.format.DateUtils
import android.view.ViewGroup
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import com.github.android.removeduplicates.BuildConfig
import com.github.android.removeduplicates.R
import com.github.android.removeduplicates.databinding.SameCallBinding
import com.github.duplicates.DuplicateItemPair
import com.github.duplicates.DuplicateViewHolder
import com.github.duplicates.SHOW_DATE_TIME
import com.github.duplicates.call.CallLogComparator.Companion.DATE
import com.github.duplicates.call.CallLogComparator.Companion.DURATION
import com.github.duplicates.call.CallLogComparator.Companion.NAME
import com.github.duplicates.call.CallLogComparator.Companion.NUMBER
import com.github.duplicates.call.CallLogComparator.Companion.TYPE
import com.github.duplicates.call.CallLogItem.Companion.ANSWERED_EXTERNALLY_TYPE
import com.github.duplicates.call.CallLogItem.Companion.BLOCKED_TYPE
import com.github.duplicates.call.CallLogItem.Companion.INCOMING_TYPE
import com.github.duplicates.call.CallLogItem.Companion.MISSED_TYPE
import com.github.duplicates.call.CallLogItem.Companion.OUTGOING_TYPE
import com.github.duplicates.call.CallLogItem.Companion.REJECTED_TYPE
import com.github.duplicates.call.CallLogItem.Companion.VOICEMAIL_TYPE

/**
 * View holder of a duplicate call.
 *
 * @author moshe.w
 */
class CallLogViewHolder(
    itemView: ViewGroup,
    binding: SameCallBinding,
    onCheckedChangeListener: OnItemCheckedChangeListener<CallLogItem>? = null
) : DuplicateViewHolder<CallLogItem>(itemView, onCheckedChangeListener) {

    private val match = binding.match

    private val checkbox1 = binding.checkbox1
    private val date1 = binding.date1
    private val duration1 = binding.duration1
    private val number1 = binding.number1
    private val type1 = binding.type1
    private val name1 = binding.name1

    private val checkbox2 = binding.checkbox2
    private val date2 = binding.date2
    private val duration2 = binding.duration2
    private val number2 = binding.number2
    private val type2 = binding.type2
    private val name2 = binding.name1

    init {
        checkbox1.setOnClickListener {
            onCheckedChangeListener?.onItemCheckedChangeListener(item1, checkbox1.isChecked)
        }
        checkbox2.setOnClickListener {
            onCheckedChangeListener?.onItemCheckedChangeListener(item2, checkbox2.isChecked)
        }
    }

    override fun bindHeader(context: Context, pair: DuplicateItemPair<CallLogItem>) {
        match.text =
            context.getString(R.string.match, percentFormatter.format(pair.match.toDouble()))
    }

    override fun bindItem1(context: Context, item: CallLogItem) {
        checkbox1.isChecked = item.isChecked
        checkbox1.text = if (BuildConfig.DEBUG) item.id.toString() else ""
        date1.text = DateUtils.formatDateTime(context, item.date, SHOW_DATE_TIME)
        duration1.text = DateUtils.formatElapsedTime(item.duration)
        type1.setImageResource(getTypeIcon(item.type))
        type1.contentDescription = context.getText(getTypeName(item.type))
        number1.text = item.number
        name1.text = item.name
    }

    override fun bindItem2(context: Context, item: CallLogItem) {
        checkbox2.isChecked = item.isChecked
        checkbox2.text = if (BuildConfig.DEBUG) item.id.toString() else ""
        date2.text = DateUtils.formatDateTime(context, item.date, SHOW_DATE_TIME)
        duration2.text = DateUtils.formatElapsedTime(item.duration)
        type2.setImageResource(getTypeIcon(item.type))
        type2.contentDescription = context.getText(getTypeName(item.type))
        number2.text = item.number
        name2.text = item.name
    }

    override fun bindDifference(context: Context, pair: DuplicateItemPair<CallLogItem>) {
        val difference = pair.difference

        bindDifference(date1, date2, difference[DATE])
        bindDifference(duration1, duration2, difference[DURATION])
        bindDifference(type1, type2, difference[TYPE])
        bindDifference(number1, number2, difference[NUMBER])
        bindDifference(name1, name2, difference[NAME])
    }

    @DrawableRes
    private fun getTypeIcon(type: Int): Int {
        return when (type) {
            INCOMING_TYPE -> R.drawable.ic_call_received
            OUTGOING_TYPE -> R.drawable.ic_call_made
            MISSED_TYPE -> R.drawable.ic_call_missed
            VOICEMAIL_TYPE -> R.drawable.ic_voicemail
            REJECTED_TYPE -> R.drawable.ic_cancel
            BLOCKED_TYPE -> R.drawable.ic_block
            ANSWERED_EXTERNALLY_TYPE -> R.drawable.ic_devices_other
            else -> R.drawable.ic_call
        }
    }

    @StringRes
    private fun getTypeName(type: Int): Int {
        return when (type) {
            INCOMING_TYPE -> R.string.call_type_incoming
            OUTGOING_TYPE -> R.string.call_type_outgoing
            MISSED_TYPE -> R.string.call_type_missed
            VOICEMAIL_TYPE -> R.string.call_type_voicemail
            REJECTED_TYPE -> R.string.call_type_rejected
            BLOCKED_TYPE -> R.string.call_type_blocked
            ANSWERED_EXTERNALLY_TYPE -> R.string.call_type_external
            else -> R.string.call_type_other
        }
    }
}
